viewing_party/party.py

```diff
@@ -1,23 +1,212 @@
+    for movie in user_data["watched"]: 
+        set_user_watched_movies.add(movie["title"])
+    
+    return set_user_watched_movies
+
+
+def friends_wateched_movies(user_data: dict) -> set:
+    set_friends_watched_movies = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            set_friends_watched_movies.add(movie["title"])
+    
+    return set_friends_watched_movies
+
+
+def remove_duplicate_values(list_of_values: list) -> list:
+    seen = set()
+    unique_values = []
+    for item in list_of_values:
+        t = tuple(item.items())
+        if t not in seen:
+            seen.add(t)
+            unique_values.append(item)
+            
+    return unique_values
+
+
+def get_unique_watched(user_data: dict) -> int:
+    user_watched_movies = user_wateched_movies(user_data)
+    friends_watched_movies = friends_wateched_movies(user_data)
+    user_unique_watched_movies = []
+    
+    set_user_unique_titles = user_watched_movies.difference(friends_watched_movies) # Set of movies(titles) only watched by user
+
+    for movie in user_data["watched"]:
+        if movie["title"] in set_user_unique_titles:
+            user_unique_watched_movies.append(movie)
+            
+    return user_unique_watched_movies
+
+
+def get_friends_unique_watched(user_data: dict) -> list: 
+    user_watched_movies = user_wateched_movies(user_data)
+    friends_watched_movies = friends_wateched_movies(user_data)
+    unique_recomended_movies = []
+     
+    set_friends_unique_movie_titles = friends_watched_movies.difference(user_watched_movies) #  Set of movies only watched by friends but not user
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in set_friends_unique_movie_titles:
+                if movie["title"] not in unique_recomended_movies:
+                    unique_recomended_movies.append(movie)
+    
+    unique_recomended_movies = remove_duplicate_values(unique_recomended_movies)
+
+    return unique_recomended_movies
+          
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data: dict) -> dict:
+    user_movies_not_watched = get_friends_unique_watched(user_data) 
+    user_subscriptions = user_data["subscriptions"]
+    recommended_movies = []
+
+    for movie in user_movies_not_watched:
+        if movie["host"] in user_subscriptions:
+            recommended_movies.append(movie)
+            
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def is_empy(user_data: dict) -> list:
+    is_empty = list()  
+
+    for friend in user_data["friends"]:
+        if friend["watched"]:
+            is_empty.append(False)
+        is_empty.append(True)
+    
+    return is_empty

```

Comment:
Ooo is the intention of this function to check if any of the friends have an empty watched list?

Reply:
Yes, I thought I needed it because I didn't know how to check every single friend watched list...